### PR TITLE
[Feature] Add HTML fuzzy hashing for structural similarity matching

### PR DESCRIPTION
**Pull request by vstakhov**
# HTML Fuzzy Hashing for Structural Similarity Matching

## Overview

This PR implements fuzzy hashing for HTML content, enabling efficient matching of email messages by HTML structure regardless of text content variations.

## Motivation

Current fuzzy hashing in Rspamd works well for text content but doesn't capture HTML structural similarity. Many legitimate emails (newsletters, notifications) and spam campaigns share common HTML templates with varying text content. This feature enables:

1. **Brand Protection**: Detect phishing emails that copy legitimate HTML structure but use different CTA domains
2. **Spam Campaigns**: Group similar spam emails by structure even when text is personalized
3. **Template Detection**: Identify emails from the same template for whitelisting/special handling
4. **Fuzzy Matching**: Find similar HTML structures with tolerance to minor changes

### Multi-Layer Approach

HTML fuzzy hash consists of 4 components:

1. **Direct Hash (exact matching)**: `blake2b(all_tokens)` - like text parts, prevents false positives
2. **Structure Shingles (fuzzy matching)**: Sliding window over DOM tag sequence
3. **CTA Domains Hash**: Extracted using existing `url_button_weights` from lua_cta
4. **Additional Hashes**: All domains (top-10), statistical features (bucketed)

### Tokenization

Each HTML tag becomes a token: `tagname[.class][@domain]`

Examples:
- `a.button@facebook.com` - button link to Facebook
- `div.header` - div with header class
- `img@cdn.example.com` - image from CDN
- `p` - plain paragraph

**Normalization:**
- Tracking classes filtered (utm, analytics, guid, etc.)
- Dynamic classes normalized (digits, UUIDs removed)
- Numeric attributes bucketed (width, height)

### Similarity Comparison

```
similarity = 0.50 × structure_similarity
           + 0.30 × cta_domains_match    # Critical for phishing!
           + 0.15 × all_domains_similarity
           + 0.05 × features_similarity
```

**Phishing Protection**: If CTA domains differ, similarity is penalized heavily (×0.3).

### Example

```
Legitimate email from Facebook:
  Tokens: div → div.header → a@facebook.com → img@fbcdn.net → p
  Direct hash: 0x1234...
  CTA domain: facebook.com
  
Phishing email:
  Tokens: div → div.header → a@evil.com → img@fbcdn.net → p
  Direct hash: 0xABCD...  ← Different!
  CTA domain: evil.com    ← Different!
  Structure similarity: 80% but CTA penalty → overall: 24% ← LOW!
```

### C API

```c
#include "shingles.h"

struct rspamd_html_shingle *html_sgl = rspamd_shingles_from_html(
    html_content,  // void* to html_content
    key,           // 16-byte key
    pool,
    rspamd_shingles_default_filter,
    NULL,
    RSPAMD_SHINGLES_MUMHASH
);

double similarity = rspamd_html_shingles_compare(html_sgl1, html_sgl2);
```

### Lua API

```lua
local text_parts = task:get_text_parts()
for _, part in ipairs(text_parts) do
  if part:is_html() then
    local digest, shingles = part:get_html_fuzzy_hashes(task:get_mempool())
    
    if digest and shingles then
      -- digest: hex string (direct hash for exact matching)
      -- shingles[1..32]: structure shingles
      -- shingles.cta_domains_hash: CTA domains
      -- shingles.all_domains_hash: all domains
      -- shingles.features_hash: statistics
      -- shingles.tags_count, links_count, dom_depth: metadata
    end
  end
end
```

### With Existing Code

- **CTA Detection**: Uses `rspamd_html_url_button_weight()` and `url_button_weights` from html_content
- **lua_cta Integration**: Leverages existing CTA analysis infrastructure
- **URL Processing**: Uses existing `rspamd_url` and `rspamd_url_find_tld()`
- **HTML Parser**: Works directly with `html_content->all_tags`

### With Fuzzy Storage

- Structure shingles compatible with `rspamd_fuzzy_shingle_cmd`
- Direct hash can be used as primary key
- Additional hashes (CTA, domains, features) can be stored as extensions
- Standard fuzzy_check protocol can be used

## Performance

- **Time**: O(n) where n = number of HTML tags, typically <1ms for 100-200 tags
- **Memory**: 
  - Temporary: ~1KB per 100 tags during generation
  - Final structure: 304 bytes (32×8 shingles + 64 direct_hash + 3×8 hashes + metadata)
- **Algorithm**: MurmurHash for speed, sliding window for tolerance

## Testing

Example test script provided in `test_html_fuzzy_hashes.lua` showing:
- Basic hash generation
- Similarity comparison
- Phishing detection scenario
- Integration examples

## Files Changed

- `src/libutil/shingles.h`: Add `rspamd_html_shingle` structure and API declarations
- `src/libutil/shingles.c`: Implement HTML fuzzy hashing (~540 lines)
- `src/lua/lua_mimepart.c`: Add `text_part:get_html_fuzzy_hashes()` Lua method

## Future Work

This is a **draft PR** for initial review. Future enhancements:

1. Integration with `fuzzy_check.c` module
2. Configuration options in `fuzzy_check.conf`:
   - `html_shingles = true/false`
   - `min_html_tags = 10`
   - `html_similarity_threshold = 0.7`
3. Rules for phishing detection based on HTML similarity
4. Separate fuzzy storage type/flag for HTML hashes
5. Performance benchmarks on real email corpus
6. Unit tests for edge cases

## Questions for Review

1. Is the multi-layer approach (direct_hash + shingles + CTA + domains + features) reasonable?
2. Should CTA domain weight (30%) be configurable?
3. Should we add more normalization (e.g., for CSS classes, inline styles)?
4. Integration with fuzzy_check: separate command or extend existing?
5. Storage: should HTML hashes be stored separately or mixed with text hashes?

### CODE REVIEW

**Comment by vstakhov** [comment]
@cursor review

**Comment by vstakhov** [comment]
@cursor review

**Comment by vstakhov** [comment]
@cursor review

File: src/lua/lua_mimepart.c [hunk 1]
```diff
@@ -1317,6 +1326,100 @@ lua_textpart_get_fuzzy_hashes(lua_State *L)
...
+		return 2;
+	}
+
+	/* Generate key */
+	rspamd_cryptobox_hash(key, "rspamd", strlen("rspamd"), NULL, 0);
+
+	/* Generate HTML shingles */
+	html_sgl = rspamd_shingles_from_html(part->html, key, pool,
+										 rspamd_shingles_default_filter,
+										 NULL, RSPAMD_SHINGLES_MUMHASH);
```
[the remark this answers is not in the document]
Comment (vstakhov) [reply src/lua/lua_mimepart.c:1371]: False positive, this function is properly declared, it is just extern "C".

File: src/libutil/shingles_html.cxx [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+	structure_sim = (double) common / (double) RSPAMD_SHINGLE_SIZE;
+
+	/* 2. CTA domains MUST match (critical for phishing) */
+	if (a->cta_domains_hash != 0 && b->cta_domains_hash != 0) {
+		cta_sim = (a->cta_domains_hash == b->cta_domains_hash) ? 1.0 : 0.0;
+
+		/* If CTA domains differ completely, likely phishing -> penalize heavily */
+		if (cta_sim < 0.5) {
+			return structure_sim * 0.3; /* Heavy penalty */
+		}
```
[the remark this answers is not in the document]
Comment (vstakhov) [reply src/libutil/shingles_html.cxx:518]: Nah, it's also ok.

File: src/libutil/shingles_html.cxx [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+						  rspamd_mempool_t *pool,
+						  rspamd_shingles_filter filter,
+						  gpointer filterd,
+						  enum rspamd_shingle_alg alg)
+{
+	if (!html_content) {
+		return nullptr;
+	}
+
+	auto *hc_ptr = html_content::from_ptr(html_content);
```
[the remark this answers is not in the document]
Comment (vstakhov) [reply src/libutil/shingles_html.cxx:435]: Not sure, looks ok.

**Comment by vstakhov** [comment]
@cursor review

File: src/libutil/shingles_html.cxx [hunk 4]
```diff
@@ -0,0 +1,540 @@
...
+
+	if (!hc_ptr) {
+		return nullptr;
+	}
```
[the remark this answers is not in the document]
Comment (vstakhov) [reply src/libutil/shingles_html.cxx:439]: huh, it is existing, false positive
